File: clients/android/app/src/main/java/network/buildit/modules/training/integration/TrainingCRMIntegration.kt
package network.buildit.modules.training.integration

import android.util.Log
import kotlinx.coroutines.flow.first
import network.buildit.modules.training.domain.model.*
import network.buildit.modules.training.domain.repository.TrainingRepository
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Contact training information for CRM display.
 */
data class ContactTrainingInfo(
    val contactId: String,
    val pubkey: String,
    val enrolledCourses: Int,
    val completedCourses: Int,
    val certifications: List<Certification>,
    val certificationsExpiring: List<Certification>,
    val totalTimeSpentHours: Float,
    val lastActivity: Long?
)

/**
 * Training requirement for volunteer roles.
 */
data class TrainingRequirement(
    val courseId: String,
    val courseName: String,
    val required: Boolean,
    val currentlyMet: Boolean,
    val certificationExpired: Boolean = false,
    val certificationExpiresAt: Long? = null
)

/**
 * Integration between Training module and CRM module.
 * Links certifications to contacts and tracks training status.
 */
@Singleton
class TrainingCRMIntegration @Inject constructor(
    private val trainingRepository: TrainingRepository
    // In a real implementation, would inject CRMManager or DatabaseManager
) {
    companion object {
        private const val TAG = "TrainingCRMIntegration"
    }

    /**
     * Adds a certification to a CRM contact record.
     *
     * @param contactId The CRM contact ID
     * @param certification The certification to add
     */
    suspend fun addCertificationToContact(
        contactId: String,
        certification: Certification
    ) {
        Log.i(TAG, "Adding certification ${certification.id} to contact $contactId")

        // In a real implementation, this would:
        // 1. Get the CRM contact record
        // 2. Update custom fields with certification info
        // 3. Save the updated contact

        // Would use DatabaseManager or CRMManager to update contact
    }

    /**
     * Gets training status for a CRM contact.
     *
     * @param contactId The CRM contact ID
     * @param pubkey The contact's public key
     * @return Training information for the contact
     */
    suspend fun getContactTrainingStatus(
        contactId: String,
        pubkey: String
    ): ContactTrainingInfo {
        val status = trainingRepository.getUserTrainingStatus(pubkey)
        val certifications = trainingRepository.getCertifications(pubkey).first()

        val now = System.currentTimeMillis() / 1000
        val thirtyDays = 30 * 24 * 60 * 60L

        val validCerts = certifications.filter { it.isValid }
        val expiringSoon = validCerts.filter { cert ->
            cert.expiresAt != null &&
                    cert.expiresAt > now &&
                    cert.expiresAt < now + thirtyDays
        }

        return ContactTrainingInfo(
            contactId = contactId,
            pubkey = pubkey,
            enrolledCourses = status.coursesEnrolled,
            completedCourses = status.coursesCompleted,
            certifications = validCerts,
            certificationsExpiring = expiringSoon,
            totalTimeSpentHours = status.totalTimeSpentHours,
            lastActivity = status.lastActivity
        )
    }

    /**
     * Filters contacts by certification status.
     *
     * @param courseId The course ID
     * @param includeExpired Whether to include expired certifications
     * @return List of pubkeys with the certification
     */
    suspend fun filterContactsByCertification(
        courseId: String,
        includeExpired: Boolean = false
    ): List<String> {
        val certifications = trainingRepository.getCertificationsForCourse(courseId).first()
        val now = System.currentTimeMillis() / 1000

        return certifications
            .filter { cert ->
                !cert.isRevoked &&
                        (includeExpired || cert.expiresAt == null || cert.expiresAt > now)
            }
            .map { it.pubkey }
            .distinct()
    }

    /**
     * Checks training requirements for a contact.
     *
     * @param pubkey The contact's public key
     * @param requiredCourseIds List of required course IDs
     * @return Whether requirements are met and detailed status
     */
    suspend fun checkTrainingRequirements(
        pubkey: String,
        requiredCourseIds: List<String>
    ): TrainingRequirementsResult {
        val certifications = trainingRepository.getCertifications(pubkey).first()
        val now = System.currentTimeMillis() / 1000

        val requirements = mutableListOf<TrainingRequirement>()
        var allMet = true

        for (courseId in requiredCourseIds) {
            val course = trainingRepository.getCourse(courseId).first()
            val cert = certifications.find { it.courseId == courseId && !it.isRevoked }

            val expired = cert?.expiresAt?.let { it < now } ?: false
            val met = cert != null && !expired

            if (!met) {
                allMet = false
            }

            requirements.add(
                TrainingRequirement(
                    courseId = courseId,
                    courseName = course?.title ?: "Unknown Course",
                    required = true,
                    currentlyMet = met,
                    certificationExpired = expired,
                    certificationExpiresAt = cert?.expiresAt
                )
            )
        }

        return TrainingRequirementsResult(
            met = allMet,
            requirements = requirements
        )
    }

    /**
     * Gets contacts with certifications expiring within a threshold.
     *
     * @param daysThreshold Number of days to look ahead
     * @return List of contacts with expiring certifications
     */
    suspend fun getContactsWithExpiringCertifications(
        daysThreshold: Int = 30
    ): List<ExpiringCertificationInfo> {
        val courses = trainingRepository.getCourses().first()
        val courseMap = courses.associateBy { it.id }

        val now = System.currentTimeMillis() / 1000
        val threshold = daysThreshold * 24 * 60 * 60L

        val results = mutableListOf<ExpiringCertificationInfo>()

        // Would need to iterate through all certifications
        // In a real implementation, this would be a more efficient database query
        courses.forEach { course ->
            val certs = trainingRepository.getCertificationsForCourse(course.id).first()
            val expiring = certs.filter { cert ->
                cert.expiresAt != null &&
                        cert.expiresAt > now &&
                        cert.expiresAt < now + threshold &&
                        !cert.isRevoked
            }

            expiring.forEach { cert ->
                results.add(
                    ExpiringCertificationInfo(
                        pubkey = cert.pubkey,
                        certification = cert,
                        course = course,
                        daysRemaining = ((cert.expiresAt!! - now) / 86400).toInt()
                    )
                )
            }
        }

        return results.sortedBy { it.daysRemaining }
    }

    /**
     * Sends certification expiry reminders.
     *
     * @param pubkeys List of public keys to notify
     * @param courseId The course ID
     * @param message Reminder message
     */
    suspend fun sendCertificationReminders(
        pubkeys: List<String>,
        courseId: String,
        message: String
    ) {
        Log.i(TAG, "Sending certification reminders to ${pubkeys.size} contacts for course $courseId")

        // In a real implementation, this would:
        // 1. Send push notifications
        // 2. Send in-app messages
        // 3. Optionally send emails if configured
    }

    /**
     * Syncs certification data to CRM.
     * Called when certifications are earned or revoked.
     *
     * @param certification The certification to sync
     */
    suspend fun syncCertificationToCRM(certification: Certification) {
        val course = trainingRepository.getCourse(certification.courseId).first()

        Log.i(TAG, "Syncing certification ${certification.id} (${course?.title}) to CRM for ${certification.pubkey}")

        // In a real implementation, this would:
        // 1. Find the CRM contact by pubkey
        // 2. Update certification-related custom fields
        // 3. Create activity record for the certification

        // Fields to update:
        // - Last certification date
        // - Certification expiry date
        // - Course name
        // - Verification code
        // - Certification status
    }

    /**
     * Checks if CRM module is available.
     */
    fun isCRMModuleAvailable(): Boolean {
        // In a real implementation, would check if CRM module is enabled
        return true // Placeholder
    }
}

/**
 * Result of checking training requirements.
 */
data class TrainingRequirementsResult(
    val met: Boolean,
    val requirements: List<TrainingRequirement>
)

/**
 * Information about an expiring certification.
 */
data class ExpiringCertificationInfo(
    val pubkey: String,
    val certification: Certification,
    val course: Course,
    val daysRemaining: Int
)

File: clients/ios/BuildIt/Modules/Training/Integrations/TrainingCRMIntegration.swift
// TrainingCRMIntegration.swift
// BuildIt - Decentralized Mesh Communication
//
// Integration between Training and CRM modules for linking certifications to contacts.

import Foundation
import os.log

/// Contact training information
public struct ContactTrainingInfo: Sendable {
    public var contactId: String
    public var pubkey: String
    public var enrolledCourses: Int
    public var completedCourses: Int
    public var certifications: [Certification]
    public var certificationsExpiring: [Certification]
    public var totalTimeSpent: Int // hours
    public var lastActivity: Date?

    public init(
        contactId: String,
        pubkey: String,
        enrolledCourses: Int = 0,
        completedCourses: Int = 0,
        certifications: [Certification] = [],
        certificationsExpiring: [Certification] = [],
        totalTimeSpent: Int = 0,
        lastActivity: Date? = nil
    ) {
        self.contactId = contactId
        self.pubkey = pubkey
        self.enrolledCourses = enrolledCourses
        self.completedCourses = completedCourses
        self.certifications = certifications
        self.certificationsExpiring = certificationsExpiring
        self.totalTimeSpent = totalTimeSpent
        self.lastActivity = lastActivity
    }
}

/// Training requirement for volunteer roles
public struct TrainingRequirement: Sendable {
    public var courseId: String
    public var courseName: String
    public var required: Bool
    public var currentlyMet: Bool
    public var certificationExpired: Bool?
    public var certificationExpiresAt: Date?

    public init(
        courseId: String,
        courseName: String,
        required: Bool,
        currentlyMet: Bool,
        certificationExpired: Bool? = nil,
        certificationExpiresAt: Date? = nil
    ) {
        self.courseId = courseId
        self.courseName = courseName
        self.required = required
        self.currentlyMet = currentlyMet
        self.certificationExpired = certificationExpired
        self.certificationExpiresAt = certificationExpiresAt
    }
}

/// Training requirements check result
public struct TrainingRequirementsResult: Sendable {
    public var met: Bool
    public var requirements: [TrainingRequirement]

    public init(met: Bool, requirements: [TrainingRequirement]) {
        self.met = met
        self.requirements = requirements
    }
}

/// Contact with expiring certification
public struct ExpiringCertificationInfo: Sendable {
    public var pubkey: String
    public var certification: Certification
    public var course: Course?

    public init(pubkey: String, certification: Certification, course: Course? = nil) {
        self.pubkey = pubkey
        self.certification = certification
        self.course = course
    }
}

/// Integration between Training and CRM modules
@MainActor
public class TrainingCRMIntegration {
    // MARK: - Properties

    private let manager: TrainingManager
    private let logger = Logger(subsystem: "com.buildit", category: "TrainingCRMIntegration")

    // MARK: - Initialization

    public init(manager: TrainingManager) {
        self.manager = manager
    }

    // MARK: - Certification Management

    /// Add certification to a CRM contact record
    public func addCertificationToContact(
        contactId: String,
        certification: Certification
    ) async throws {
        // In a real implementation, this would update the CRM contact record
        // with the certification information
        logger.info("Adding certification \(certification.id) to contact \(contactId)")

        // Would store in CRM custom fields
        // This would interact with the database module/CRM module
    }

    /// Get training status for a CRM contact
    public func getContactTrainingStatus(
        contactId: String,
        pubkey: String
    ) async throws -> ContactTrainingInfo {
        let status = try await manager.getUserTrainingStatus(pubkey: pubkey)
        let certifications = try await manager.listCertifications(pubkey: pubkey)

        let now = Date()
        let thirtyDays = TimeInterval(30 * 24 * 60 * 60)
        let validCertifications = certifications.filter { $0.isValid }
        let expiringSoon = certifications.filter { cert in
            guard let expiresAt = cert.expiresAt else { return false }
            return expiresAt > now && expiresAt < now.addingTimeInterval(thirtyDays) && cert.revokedAt == nil
        }

        return ContactTrainingInfo(
            contactId: contactId,
            pubkey: pubkey,
            enrolledCourses: status.coursesEnrolled,
            completedCourses: status.coursesCompleted,
            certifications: validCertifications,
            certificationsExpiring: expiringSoon,
            totalTimeSpent: status.totalTimeSpent,
            lastActivity: status.lastActivity
        )
    }

    /// Filter CRM contacts by certification
    public func filterContactsByCertification(
        courseId: String,
        includeExpired: Bool = false
    ) async throws -> [String] {
        let certifications = try await manager.listCertifications()

        let now = Date()
        let filtered = certifications.filter { cert in
            if cert.courseId != courseId { return false }
            if cert.revokedAt != nil { return false }
            if !includeExpired {
                if let expiresAt = cert.expiresAt, expiresAt < now { return false }
            }
            return true
        }

        // Return unique pubkeys
        return Array(Set(filtered.map { $0.pubkey }))
    }

    /// Check training requirements for a contact
    public func checkTrainingRequirements(
        pubkey: String,
        requiredCourseIds: [String]
    ) async throws -> TrainingRequirementsResult {
        let certifications = try await manager.listCertifications(pubkey: pubkey)
        let now = Date()

        var requirements: [TrainingRequirement] = []
        var allMet = true

        for courseId in requiredCourseIds {
            let course = try await manager.getCourse(id: courseId)
            let cert = certifications.first { $0.courseId == courseId && $0.revokedAt == nil }

            let expired: Bool
            if let expiresAt = cert?.expiresAt {
                expired = expiresAt < now
            } else {
                expired = false
            }

            let met = cert != nil && !expired

            if !met {
                allMet = false
            }

            requirements.append(TrainingRequirement(
                courseId: courseId,
                courseName: course?.title ?? "Unknown Course",
                required: true,
                currentlyMet: met,
                certificationExpired: expired,
                certificationExpiresAt: cert?.expiresAt
            ))
        }

        return TrainingRequirementsResult(met: allMet, requirements: requirements)
    }

    /// Get contacts with expiring certifications
    public func getContactsWithExpiringCertifications(
        daysThreshold: Int = 30
    ) async throws -> [ExpiringCertificationInfo] {
        let certifications = try await manager.listCertifications()

        let now = Date()
        let threshold = TimeInterval(daysThreshold * 24 * 60 * 60)

        let expiring = certifications.filter { cert in
            guard let expiresAt = cert.expiresAt else { return false }
            return expiresAt > now && expiresAt < now.addingTimeInterval(threshold) && cert.revokedAt == nil
        }

        var results: [ExpiringCertificationInfo] = []
        for cert in expiring {
            let course = try await manager.getCourse(id: cert.courseId)
            results.append(ExpiringCertificationInfo(
                pubkey: cert.pubkey,
                certification: cert,
                course: course
            ))
        }

        return results
    }

    /// Send certification reminders to contacts
    public func sendCertificationReminders(
        pubkeys: [String],
        courseId: String,
        message: String
    ) async throws {
        // In a real implementation, this would send notifications via the messaging module
        logger.info("Sending certification reminders to \(pubkeys.count) contacts for course \(courseId)")

        // Would integrate with messaging/notifications module
    }

    /// Check if CRM module is available
    public func isCRMModuleAvailable() async -> Bool {
        // In a real implementation, would check if CRM module is enabled
        return true
    }

    /// Sync certification data to CRM
    /// Called when certifications are earned/revoked
    public func syncCertificationToCRM(certification: Certification) async throws {
        let course = try await manager.getCourse(id: certification.courseId)

        logger.info("Syncing certification \(certification.id) (\(course?.title ?? "Unknown")) to CRM for \(certification.pubkey)")

        // Would update CRM custom fields with:
        // - Certification earned date
        // - Certification expiry date
        // - Course name
        // - Verification code
    }

    /// Get all certified contacts for a course
    public func getCertifiedContacts(courseId: String) async throws -> [String] {
        try await filterContactsByCertification(courseId: courseId, includeExpired: false)
    }

    /// Get training completion rate for a group
    public func getGroupTrainingCompletion(
        groupId: String,
        courseId: String
    ) async throws -> (enrolled: Int, completed: Int, certified: Int) {
        // In a real implementation, this would query CRM and training data
        // to get completion statistics for all contacts in a group

        logger.info("Getting training completion for group \(groupId), course \(courseId)")

        return (enrolled: 0, completed: 0, certified: 0)
    }

    /// Bulk enroll contacts in a course
    public func bulkEnrollContacts(
        contactPubkeys: [String],
        courseId: String
    ) async throws {
        // In a real implementation, would create enrollments for all specified contacts
        logger.info("Bulk enrolling \(contactPubkeys.count) contacts in course \(courseId)")
    }

    /// Generate training report for CRM
    public func generateTrainingReport(
        groupId: String?,
        startDate: Date,
        endDate: Date
    ) async throws -> TrainingReport {
        // In a real implementation, would aggregate training data for the period

        return TrainingReport(
            groupId: groupId,
            startDate: startDate,
            endDate: endDate,
            totalEnrollments: 0,
            totalCompletions: 0,
            certificationIssued: 0,
            certificationsRevoked: 0,
            averageCompletionTime: 0,
            courseStats: []
        )
    }
}

/// Training report data
public struct TrainingReport: Sendable {
    public var groupId: String?
    public var startDate: Date
    public var endDate: Date
    public var totalEnrollments: Int
    public var totalCompletions: Int
    public var certificationIssued: Int
    public var certificationsRevoked: Int
    public var averageCompletionTime: Double // hours
    public var courseStats: [CourseReportStats]

    public init(
        groupId: String? = nil,
        startDate: Date,
        endDate: Date,
        totalEnrollments: Int,
        totalCompletions: Int,
        certificationIssued: Int,
        certificationsRevoked: Int,
        averageCompletionTime: Double,
        courseStats: [CourseReportStats]
    ) {
        self.groupId = groupId
        self.startDate = startDate
        self.endDate = endDate
        self.totalEnrollments = totalEnrollments
        self.totalCompletions = totalCompletions
        self.certificationIssued = certificationIssued
        self.certificationsRevoked = certificationsRevoked
        self.averageCompletionTime = averageCompletionTime
        self.courseStats = courseStats
    }
}

/// Per-course statistics in a report
public struct CourseReportStats: Sendable {
    public var courseId: String
    public var courseName: String
    public var enrollments: Int
    public var completions: Int
    public var certifications: Int

    public init(
        courseId: String,
        courseName: String,
        enrollments: Int,
        completions: Int,
        certifications: Int
    ) {
        self.courseId = courseId
        self.courseName = courseName
        self.enrollments = enrollments
        self.completions = completions
        self.certifications = certifications
    }
}
